### DISCUSSION TITLE
OAuth2 with restriction for manual user creation

### DISCUSSION THREAD

**Discussion by m-2k** [pocketbase/pocketbase#4223] [Q&A] [answered]
I use `pb.collection("users").authWithOAuth2Code(...)` but i want restrict user creation API calls `pb.collection("users").create(...)`

What to do? :)

**Comment by ganigeorgiev** [comment, marked as answer]
At the moment there is no special filter field to distinguish the regular user creation from the OAuth2 one (_I have it in my todo to consider adding a context or similar as other users also requested it for the expand and realtime apis_).

As a workaround you can use the [`OnRecordBeforeCreateRequest`](https://pocketbase.io/docs/go-event-hooks/#onrecordbeforecreaterequest) hook and return an error for non-admin requests, something like:

- Go:
    ```go
    app.OnRecordBeforeCreateRequest("users").Add(func(e *core.RecordCreateEvent) error {
        // allow for admins
        if e.HttpContext.Get(apis.ContextAdminKey) != nil {
            return nil
        }

        // disallow for everyone else
        return apis.NewForbiddenError("You are not allowed to perform this action.", nil)
    })
    ```
- JS:
    ```js
    onRecordBeforeCreateRequest((e) => {
        // allow for admins
        if (e.httpContext.get("admin")) {
            return
        }

        // disallow for everyone else
        throw new ForbiddenError("You are not allowed to perform this action.") 
    }, "users")
    ```

**Reply by m-2k** [reply]
@ganigeorgiev hey, maybe use a trick like this:
Create rule for `users` collection:
```
@request.data.*:isset = false
```
(but for multicast `.*` there are no checks `:isset` for missing data, you will have to enumerate manually; or set some field in the table as mandatory and check only on it)

**Comment by m-2k** [comment]
Before writing the post, I have seen issue https://github.com/pocketbase/pocketbase/discussions/4061, but it didn’t work on the first try.

I rechecked it now - it works (but the CREATE API rule for `users` table without text and with access for everyone). But it all doesn’t look intuitive.
I think for the authorization tables it would be logical to add, in addition to CRUD API rules, additional fields with rules or checkboxes for the AUTH API.

Thx.

**Reply by ganigeorgiev** [reply]
There are no plans for now to add more API rule fields/controls, but as mentioned earlier with the next v0.22/v0.23 release I'll consider adding a special "context" or similar filter key that could be used to identify the execution context in case a single rule is used in more than one place (or in other words, you'll be able to do something like: `@request.context = "oauth2"`).

**Reply by m-2k** [reply]
if adding a context field would make sense somewhere other than authorization, then it's probably a reasonable solution. but still unintuitive and increasing the entry threshold.

If adding additional UI fields for authorization is in principle in the plans, then I would rather wait for the option with fields than for an additional entity like context field.

**Reply by m-2k** [reply]
I think if there is an understanding of how PB works from the inside (which is not suitable for new PB users), then the additional field in the rules looks more logical. And if you look from the outside, my understanding of what is happening with the rules is as follows:
- creation rule only for POST requests api/collections/*/records
- the authorization method has the side effect of creating a record, but there are no rules for these requests in the UI

I don't oblige you to anything, just take note
